exquisite_corpus/spm_naive_language_model.py

```diff
@@ -0,0 +1,428 @@
+        self,
+        model=None,
+        device=torch.device("cuda:0"),
+        **kwargs
+    ):
+        if model is None:
+            self.model = LanguageModel(**kwargs)
+        else:
+            self.model = model
+        self.device = device
+
+    @property
+    def device(self):
+        return self._device
+
+    @device.setter
+    def device(self, new_device):
+        if isinstance(new_device, str):
+            new_device = torch.device(new_device)
+        if new_device != torch.device("cpu") and not torch.cuda.is_available():
+            print("Warning:  cuda not available so setting device to CPU.")
+            self._device = torch.device("cpu")
+        else:
+            self._device = new_device
+        self.model.to(self._device)
+
+    def collate_batch_with_labels(self, raw_batch):
+        """
+        Take a raw batch (as returned by a DataLoader and SpmIdsBatchSampler),
+        collate it, and split off labels.
+        """
+        data = torch.LongTensor(np.vstack([x[:-1] for x in raw_batch]))
+        labels = torch.LongTensor(np.vstack([x[1:] for x in raw_batch]))
+        return data, labels
+
+    def train(
+        self,
+        train_dataset_path,
+        validation_dataset_path=None,
+        save_path=None,
+        min_length=2,
+        n_epochs=np.inf,
+        n_batches_between_saves=1000,
+        n_batches_between_messages=50,
+        n_batches_between_validations=25000,
+        batch_size=256,
+        n_data_loader_workers=0,
+        lr=0.05,
+        device=None,
+    ):
+        if device is not None:
+            self.device = device
+
+        self.model.train()
+        if min_length < 2:
+            raise ValueError("Cannot train with sentences shorter than 2 tokens.")
+        train_data = SpmIdsDataset(train_dataset_path, min_length=min_length)
+        sampler = SpmIdsBatchSampler(train_data, batch_size=batch_size)
+        pin_memory = self.device != torch.device("cpu")
+        data_loader = DataLoader(
+            train_data,
+            batch_sampler=sampler,
+            pin_memory=pin_memory,
+            num_workers=n_data_loader_workers,
+            collate_fn=self.collate_batch_with_labels,
+        )
+        if validation_dataset_path is None:
+            validation_data = None
+        else:
+            validation_data = SpmIdsDataset(
+                validation_dataset_path, min_length=min_length
+            )
+            validation_sampler = SpmIdsBatchSampler(
+                validation_data, batch_size=batch_size, randomize=False
+            )
+            validation_data_loader = DataLoader(
+                validation_data,
+                batch_sampler=validation_sampler,
+                pin_memory=pin_memory,
+                num_workers=n_data_loader_workers,
+                collate_fn=self.collate_batch_with_labels,
+            )
+
+        optimizer = torch.optim.SGD(self.model.parameters(), lr=lr)
+        loss_function = LossFunction()
```

Comment:
There's a loss function just called LossFunction? This is nearly ungoogleable. What does it do?

Reply:
Oh, I missed that LossFunction is specific code implemented here. Still, a more specific name would help, as this sounds like it should be a superclass of all loss functions.

Reply:
This now has a more specific name.